src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs

```diff
@@ -1882,6 +1882,94 @@ internal override IList<PSModuleInfo> ImportModulesUsingWinCompat(IEnumerable<st
         {
             IList<PSModuleInfo> moduleProxyList = new List<PSModuleInfo>();
 #if !UNIX
+            // the ModuleDeny list is cached in PowerShellConfig object
+            string[] moduleDenyList = System.Management.Automation.Configuration.PowerShellConfig.Instance.GetWindowsPowerShellCompatibilityModuleDenyList();
+
+            // one of the two parameters can be passed: either ModuleNames (most of the time) or ModuleSpecifications (they are used in different parameter sets)
+            List<string> filteredModuleNames = null;
+            if (moduleNames != null)
+            {
+                if ((moduleDenyList == null) || (moduleDenyList.Length == 0))
+                {
+                    filteredModuleNames = new List<string>(moduleNames);
+                }
+                else
+                {
+                    filteredModuleNames = new List<string>();
+                    foreach(var moduleName in moduleNames)
+                    {
+                        // moduleName can be just a module name and it also can be a full path to psd1 from which we need to extract the module name
+                        var exactModuleName = System.IO.Path.GetFileNameWithoutExtension(moduleName);
+                        bool match = false;
+                        foreach(var deniedModuleName in moduleDenyList)
+                        {
+                            // use case-insensitive module name comparison
+                            match = exactModuleName.Equals(deniedModuleName, StringComparison.InvariantCultureIgnoreCase);
+                            if (match)
+                            {
+                                string errorMessage = string.Format(CultureInfo.InvariantCulture, Modules.WinCompatModuleInDenyList, exactModuleName);
+                                InvalidOperationException exception = new InvalidOperationException(errorMessage);
+                                ErrorRecord er = new ErrorRecord(exception, "Modules_ModuleInWinCompatDenyList", ErrorCategory.ResourceUnavailable, moduleName);
+                                WriteError(er);
+                                break;
+                            }
+                        }
+
+                        if (!match)
+                        {
+                            filteredModuleNames.Add(moduleName);
+                        }
+                    }
+                }
+
+                // do not setup WinCompat resources if we have no modules to import
+                if (filteredModuleNames.Count == 0)
+                {
+                    return moduleProxyList;
+                }
+            }
+
+            List<ModuleSpecification> filteredModuleFullyQualifiedNames = null;
```

Comment:
Should we put the code below in `else` clause because ModuleNames  and ModuleSpecifications are in different parameter sets?

Code for both ModuleNames  and ModuleSpecifications is the same - can we use a generic method?

Reply:
`Should we put the code below in else...` - I had no strong oppinion on this until I looked at at the child function `ImportModule_RemotelyViaPsrpSession` - it takes both parameters and processes them both if they are not null, but it does not make them mutually exclusive, so I think we should follow the same pattern here. Parameter sets will take care of separation.

`Code for both ModuleNames and ModuleSpecifications is the same - can we use a generic method?` - that was i was going to do originally, but turns out there are differences: different object types (string vs ModuleSpecification), sting-case needs special processing cause it can be a full path, ModuleSpec-case needs special processing cause we need to compare object property and not the object itself, same for the error message formatting. I feel like templatizing this will just make this harder to read and won't save as much space, so probably not worth it.